src/coreclr/src/vm/nativeimage.cpp

```diff
@@ -0,0 +1,156 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// --------------------------------------------------------------------------------
+// NativeImage.cpp
+//
+
+// --------------------------------------------------------------------------------
+
+#include "common.h"
+#include "nativeimage.h"
+
+// --------------------------------------------------------------------------------
+// Headers
+// --------------------------------------------------------------------------------
+
+#include <shlwapi.h>
+
+BOOL AssemblyNameIndexHashTraits::Equals(LPCUTF8 a, LPCUTF8 b)
+{
+    WRAPPER_NO_CONTRACT;
+
+    SString s1;
+    SString s2;
+    s1.SetUTF8(a);
+    s2.SetUTF8(b);
+    return s1.CompareCaseInsensitive(s2) == 0;
+}
+
+AssemblyNameIndexHashTraits::count_t AssemblyNameIndexHashTraits::Hash(LPCUTF8 s)
+{
+    WRAPPER_NO_CONTRACT;
+
+    SString s1;
+    s1.SetUTF8(s);
+    return s1.HashCaseInsensitive();
+}
+
+NativeImage::NativeImage(
+    NewHolder<PEImageLayout>& peImageLayout,
+    READYTORUN_HEADER *pHeader,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator,
+    AllocMemTracker *pamTracker)
+    : m_eagerFixupsLock(CrstNativeImageEagerFixups)
+{
+    CONTRACTL
+    {
+        THROWS;
+        CONSTRUCTOR_CHECK;
+        STANDARD_VM_CHECK;
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END;
+
+    LoaderHeap *pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+    m_utf8SimpleName = nativeImageName;
+    m_peImageLayout.Assign(peImageLayout.Extract());
+    m_eagerFixupsHaveRun = false;
+
+    m_pReadyToRunInfo.Assign(new ReadyToRunInfo(/*pModule*/ NULL, peImageLayout, pHeader, /*compositeImage*/ NULL, pamTracker), /*takeOwnership*/ TRUE);
+    m_pComponentAssemblies = m_pReadyToRunInfo->FindSection(ReadyToRunSectionType::ComponentAssemblies);
+    m_componentAssemblyCount = m_pComponentAssemblies->Size / sizeof(READYTORUN_COMPONENT_ASSEMBLIES_ENTRY);
+    
+    // Check if the current module's image has native manifest metadata, otherwise the current->GetNativeAssemblyImport() asserts.
+    m_pManifestMetadata = LoadManifestMetadata();
+
+    HENUMInternal assemblyEnum;
+    HRESULT hr = m_pManifestMetadata->EnumAllInit(mdtAssemblyRef, &assemblyEnum);
+    mdAssemblyRef assemblyRef;
+    int assemblyIndex = 0;
+    while (m_pManifestMetadata->EnumNext(&assemblyEnum, &assemblyRef))
+    {
+        LPCSTR assemblyName;
+        hr = m_pManifestMetadata->GetAssemblyRefProps(assemblyRef, NULL, NULL, &assemblyName, NULL, NULL, NULL, NULL);
+        m_assemblySimpleNameToIndexMap.Add(AssemblyNameIndex(assemblyName, assemblyIndex));
+        assemblyIndex++;
+    }
+}
+
+NativeImage::~NativeImage()
+{
+    if (m_pManifestMetadata != NULL)
+    {
+        m_pManifestMetadata->Release();
+    }
+}
+
+bool NativeImage::Matches(LPCUTF8 utf8SimpleName) const
+{
+    return !_stricmp(utf8SimpleName, m_utf8SimpleName);
```

Comment:
It this is actually a filename, it would be more appropriate to conditionalize this using `FEATURE_CASE_SENSITIVE_FILESYSTEM` like we do in other places.

Reply:
Good point, fixed in 16th commit.

Reply:
The other place (`AssemblyNameIndexHashTraits::Equals()`) that does a case-insensitive string comparison of LPCUTC8 values converts this to SStrings and whatnot; maybe just use _stricmp() like here and avoid creating those temporaries altogether?

Reply:
Thanks for pointing that out. That was actually a bug, I believe <code>_stricmp</code> isn't guaranteed to perform a correct UTF8 case insensitive comparison in the general case even though the way Unicode and UTF8 are constructed it may just work in common cases. I have replaced this bit to use <code>SString</code>-based case insensitive comparison too.